bugbug/repository.py

```diff
@@ -1014,42 +1014,72 @@ def apply_stack(repo_dir, stack, branch, default_base):
     """Apply a stack of patches on a repository"""
     assert len(stack) > 0, "Empty stack"
 
-    # Start by updating the repository
-    clean(repo_dir)
-
     def has_revision(revision):
         try:
             hg.identify(revision)
             return True
         except hglib.error.CommandError:
             return False
 
-    with hglib.open(repo_dir) as hg:
+    def apply_patches(base, patches):
+
+        # Update to base revision
+        logger.info(f"Updating {repo_dir} to {base}")
+        hg.update(base, clean=True)
 
-        # Find the base revision to apply all the patches onto
-        # Use first parent from first patch if all its parents are available
-        # Otherwise fallback on tip
+        # Then apply each patch in the stack
+        try:
+            for node, patch in patches:
+                hg.import_(patches=io.BytesIO(patch.encode("utf-8")), user="bugbug")
+        except hglib.error.CommandError as e:
+            logger.warning(f"Failed to apply patch {node}: {e.err}")
+            return False
+        return True
+
+    # Find the base revision to apply all the patches onto
+    # Use first parent from first patch if all its parents are available
+    # Otherwise fallback on tip
+    def get_base():
         parents = stack[0]["parents"]
         assert len(parents) > 0, "No parents found for first patch"
         if all(map(has_revision, parents)):
-            base = parents[0]
-        else:
-            # Some repositories need to have the exact parent to apply
-            if default_base is None:
-                raise Exception("Parents are not available, cannot apply this stack")
+            return parents[0]
 
-            base = default_base
+        # Some repositories need to have the exact parent to apply
+        if default_base is None:
+            raise Exception("Parents are not available, cannot apply this stack")
 
-        # Update to base revision
-        logger.info(f"Will apply stack on {base}")
-        hg.update(base, clean=True)
+        return default_base
+
+    with hglib.open(repo_dir) as hg:
```

Comment:
As a first step, we have to run part of `clean` (reverting and stripping), otherwise when we apply a stack after another one, we might fail.